apps/web/src/app/(dashboard)/admin/users/data-table-users.tsx

```diff
@@ -0,0 +1,193 @@
+'use client';
+
+import { useEffect, useState, useTransition } from 'react';
+
+import Link from 'next/link';
+
+import { Edit, Loader } from 'lucide-react';
+
+import { useDebouncedValue } from '@documenso/lib/client-only/hooks/use-debounced-value';
+import { useUpdateSearchParams } from '@documenso/lib/client-only/hooks/use-update-search-params';
+import { Document, Role, Subscription } from '@documenso/prisma/client';
+import { Button } from '@documenso/ui/primitives/button';
+import { DataTable } from '@documenso/ui/primitives/data-table';
+import { DataTablePagination } from '@documenso/ui/primitives/data-table-pagination';
+import { Input } from '@documenso/ui/primitives/input';
+import { useToast } from '@documenso/ui/primitives/use-toast';
+
+import { search } from './fetch-users.actions';
+
+interface User {
+  id: number;
+  name: string | null;
+  email: string;
+  roles: Role[];
+  Subscription: SubscriptionLite[];
+  Document: DocumentLite[];
+}
+
+type SubscriptionLite = Pick<
+  Subscription,
+  'id' | 'status' | 'planId' | 'priceId' | 'createdAt' | 'periodEnd'
+>;
+
+type DocumentLite = Pick<Document, 'id'>;
+
+type UsersDataTableProps = {
+  perPage: number;
+  page: number;
+};
+
+export const UsersDataTable = ({ perPage, page }: UsersDataTableProps) => {
+  const { toast } = useToast();
+
+  const [isPending, startTransition] = useTransition();
+  const updateSearchParams = useUpdateSearchParams();
+  const [data, setData] = useState<User[]>([]);
+  const [searchString, setSearchString] = useState('');
+  const [totalPages, setTotalPages] = useState<number>(0);
+  const debouncedSearchString = useDebouncedValue(searchString, 500);
+
+  const onPaginationChange = (page: number, perPage: number) => {
+    startTransition(() => {
+      updateSearchParams({
+        page,
+        perPage,
+      });
+    });
+  };
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const result = await search(debouncedSearchString, page, perPage);
+        setData(result.users);
+        setTotalPages(result.totalPages);
+
+        if (result.totalPages < page) {
+          startTransition(() => {
+            updateSearchParams({
+              page: 1,
+              perPage,
+            });
+          });
+        }
+      } catch (err) {
+        throw new Error(err);
+      }
+    };
```

Comment:
Typically this would be handled in the server component one level above e.g. `admin/users/page.tsx` so within the data table we're just manipulating the search params which will cause an update.